src/blog/wasm-tail-call.md

```diff
@@ -0,0 +1,165 @@
+---
+title: 'WebAssembly Tail Calls'
+description: 'This document explains the WebAssembly Tail Calls proposal and
+demonstrates it with some examples'
+author: 'Thibaud Michaud, Thomas Lively'
+date: 2023-04-04
+tags:
+  - WebAssembly
+---
+We are shipping WebAssembly Tail Calls in V8 11.2! In this post we give a brief overview of this proposal, demonstrate an interesting use case for C++ coroutines with Emscripten, and show how V8 handles tail calls internally.
+
+
+## What is Tail Call Optimization
+
+A call is said to be in tail position if it is the last instruction executed before returning from the current function. Compilers can optimize such calls by discarding the caller frame and replacing the call with a jump.
+
+This is especially useful for recursive functions. For instance, take this C function that sums the elements of a linked list:
+
+```c
+int sum(List* list, int acc) {
+  if (list == nullptr) return acc;
+  return sum(list->next, acc + list->val);
+}
+```
+
+With a regular call, this consumes O(n) stack space: each element of the list adds a new frame on the call stack. With a long enough list, this could very quickly overflow the stack. By replacing the call with a jump, tail-call optimization effectively turns this recursive function into a loop which uses
+O(1) stack space:
+
+```c
+int sum(List* list, int acc) {
+  while (list != nullptr) {
+    acc = acc + list->val;
+    list = list->next;
+  }
+}
+```
+
+This optimization is particularly important for functional languages. They rely heavily on recursive functions, and pure ones like Haskell don’t even provide loop control structures. Any kind of custom iteration typically uses recursion one way or another. Without tail-call optimization, this would very quickly run into a stack overflow for any non-trivial program.
+
+### The WebAssembly Tail Call Proposal
+
+There are two ways to call a function in Wasm MVP: `call` and `call_indirect`.  The Tail Call proposal adds their tail-call counterparts: `return_call` and `return_call_indirect`. This means that it is the responsibility of the toolchain to actually perform tail-call optimization and emit the appropriate call kind, which gives it more control over performance and stack space usage.
+
+Let’s look at a recursive fibonacci function. The Wasm bytecode is included here in the text format for completeness, but you can find it in C++ in the next section:
```

Comment:
There is a mix of `’` and `'` used in this file. We should clean that up (and always use `'`?).